backend/pkg/config/kafka_sasl_oauth.go

```diff
@@ -10,25 +10,70 @@
 package config
 
 import (
+	"context"
+	"encoding/base64"
 	"flag"
 	"fmt"
+	"io/ioutil"
+	"net/http"
 )
 
 // KafkaSASLOAuthBearer is the config struct for the SASL OAuthBearer mechanism
 type KafkaSASLOAuthBearer struct {
-	Token string `yaml:"token"`
+	Token         string `yaml:"token"`
+	ClientId      string `yaml:"clientid"`
+	ClientSecret  string `yaml:"clientsecret"`
+	TokenEndpoint string `yaml:"tokenendpoint"`
 }
 
-// RegisterFlags registers all sensitive Kerberos settings as flag
 func (c *KafkaSASLOAuthBearer) RegisterFlags(f *flag.FlagSet) {
 	f.StringVar(&c.Token, "kafka.sasl.oauth.token", "", "OAuth Bearer Token")
+	f.StringVar(&c.ClientId, "kafka.sasl.oauth.clientid", "", "OAuth Bearer Client Id")
+	f.StringVar(&c.ClientSecret, "kafka.sasl.oauth.clientsecret", "", "OAuth Bearer Client Secret")
+	f.StringVar(&c.TokenEndpoint, "kafka.sasl.oauth.tokenendpoint", "", "OAuth Bearer Token Endpoint")
 }
 
 // Validate Kafka SASL OAuth configurations.
 func (c *KafkaSASLOAuthBearer) Validate() error {
-	if c.Token == "" {
+	if c.TokenEndpoint != "" {
+		if c.ClientId == "" || c.ClientSecret == "" {
+			return fmt.Errorf("OAuth Bearer client credentials must be set")
+		}
+	} else if c.Token == "" {
 		return fmt.Errorf("OAuth Bearer token must be set")
 	}
 
 	return nil
 }
+
+func (c *KafkaSASLOAuthBearer) AcquireToken(ctx context.Context) (string, error) {
```

Comment:
We could also use a standard library for this, but I am not sure what opinion you had on including new libraries

Reply:
What library do you have in mind? I'm mostly afraid of doing something wrong in the OAuth token process, but the code looks straight forward and given that the authority is the other side which decides whether the token is valid or not, I'm also not that afraid.